File: src/caffeine_lang/frontend/tokenizer_ffi.mjs
// Fast tokenizer primitives. gleam_stdlib's string.pop_grapheme constructs
// a fresh Intl.Segmenter().segment(s)[Symbol.iterator]() on every call and
// walks it from index 0, turning the tokenizer's per-char loop into O(N^2).
// We split by codepoint instead — which only differs from grapheme for
// combining-mark / ZWJ sequences inside string-literal or comment bodies,
// neither of which the tokenizer needs to slice grapheme-correctly (it just
// scans for ASCII terminators).

// Returns [first_codepoint_as_string, rest]. Returns ["", ""] when source is
// empty — the Gleam wrapper turns that back into Error(Nil).
export function pop_codepoint(s) {
  if (s.length === 0) return ["", ""];
  const cp = s.codePointAt(0);
  const charLen = cp > 0xffff ? 2 : 1;
  return [String.fromCodePoint(cp), s.substring(charLen)];
}

// Returns the UTF-16 code unit at index i, or -1 if i is out of bounds.
// Used by is_digit / is_letter on single-codepoint strings.
export function code_unit_at(s, i) {
  if (i < 0 || i >= s.length) return -1;
  return s.charCodeAt(i);
}

// Codeunit length — used to advance the column counter without re-walking
// the just-read token via Intl.Segmenter.
export function code_unit_length(s) {
  return s.length;
}

File: src/caffeine_lang/frontend/tokenizer_ffi.erl
%% Erlang counterparts to tokenizer_ffi.mjs. Erlang strings are UTF-8 binaries,
%% so the UTF-8 pattern match cleanly extracts one codepoint at a time.
-module(tokenizer_ffi).
-export([pop_codepoint/1, code_unit_at/2, code_unit_length/1]).

pop_codepoint(<<>>) -> {<<>>, <<>>};
pop_codepoint(<<Cp/utf8, Rest/binary>>) -> {<<Cp/utf8>>, Rest};
pop_codepoint(_) -> {<<>>, <<>>}.

code_unit_at(S, I) when is_binary(S), is_integer(I), I >= 0 ->
    case byte_size(S) > I of
        true -> binary:at(S, I);
        false -> -1
    end;
code_unit_at(_, _) -> -1.

code_unit_length(S) when is_binary(S) -> byte_size(S);
code_unit_length(_) -> 0.
